[ci] [python-package] scikit-learn compatibility tests fail: "fitting with `sample_weight` is not equivalent to fitting with removed or repeated data points."

## Description

LightGBM has a CI job that tests the Python package against nightly versions of several dependencies. That job started failing 2 days ago because of the changes in https://github.com/scikit-learn/scikit-learn/pull/29818.

> FAILED tests/python_package_test/test_sklearn.py::test_sklearn_integration[LGBMClassifier()-check_sample_weight_equivalence] - AssertionError:
>
> E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=1e-09
E           Comparing the output of LGBMClassifier.predict_proba revealed that fitting with `sample_weight` is not equivalent to fitting with removed or repeated data points.
E           Mismatched elements: 15 / 45 (33.3%)
E           Max absolute difference: 0.63584606
E           Max relative difference: 10883.68807153
E            x: array([[9.998613e-01, 6.247023e-05, 7.625781e-05],
E                  [1.923383e-05, 9.999402e-01, 4.061468e-05],
E                  [1.694735e-05, 2.947391e-05, 9.999536e-01],...
E            y: array([[9.998613e-01, 6.247023e-05, 7.625780e-05],
E                  [1.923383e-05, 9.999402e-01, 4.061468e-05],
E                  [1.694735e-05, 2.947391e-05, 9.999536e-01],...

This is only happening with the `1.6.dev0` nightlies of `scikit-learn`.

## Reproducible example

This is happening across all pull requests and builds on `master`: https://github.com/microsoft/LightGBM/actions/runs/11317668594/job/31471392246

To reproduce locally:

```shell
cmake -B build -S .
cmake --build build --target _lightgbm -j4
sh build-python.sh install --precompile

python -m pip install \
    --extra-index-url https://pypi.anaconda.org/scientific-python-nightly-wheels/simple \
    --prefer-binary \
    --pre \
    --no-deps \
        'scikit-learn>=1.6.dev0'

pytest 'tests/python_package_test/test_sklearn.py::test_sklearn_integration'
```

## Environment info

N/A

## Additional Comments

It seems `scikit-learn` is making several significant changes to these estimator checks on the way to a 1.6 release. See #6653 and the work to fix it (#6651).

--- a/.ci/conda-envs/ci-core-py38.txt
+++ b/.ci/conda-envs/ci-core-py38.txt
@@ -41,6 +41,7 @@ bokeh=3.1.*
 fsspec=2024.5.*
 msgpack-python=1.0.*
 pluggy=1.5.*
+pyparsing=3.1.4
 pytz=2024.1
 setuptools=69.5.*
 snappy=1.2.*

--- a/python-package/lightgbm/sklearn.py
+++ b/python-package/lightgbm/sklearn.py
@@ -678,6 +678,12 @@ class LGBMModel(_LGBMModelBase):
                 "check_no_attributes_set_in_init": "scikit-learn incorrectly asserts that private attributes "
                 "cannot be set in __init__: "
                 "(see https://github.com/microsoft/LightGBM/issues/2628)",
+                "check_sample_weight_equivalence": (
+                    "In LightGBM, setting a sample's weight to 0 can produce a different result than omitting the sample. "
+                    "Such samples intentionally still affect count-based measures like 'min_data_in_leaf' "
+                    "(https://github.com/microsoft/LightGBM/issues/5626#issuecomment-1712706678) and the estimated distribution "
+                    "of features for Dataset construction (see https://github.com/microsoft/LightGBM/issues/5553)."
+                ),
             },
         }
 

